File: 236-lowest-common-ancestor-of-a-binary-tree.java
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {       
        if(root == null ) { return null; }
        
        Map<TreeNode, TreeNode> childParentMap = new HashMap<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        while(!stack.isEmpty()) {
            if(childParentMap.containsKey(p) && childParentMap.containsKey(q)) {
                break;
            }
            TreeNode node = stack.pop();
            if(node.right != null) {
                stack.push(node.right);
                childParentMap.put(node.right, node);
            }
            
            if(node.left != null) {
                stack.push(node.left);
                childParentMap.put(node.left, node);
            }
        }
        
        Set<TreeNode> ancestorsOfP = new HashSet<>();
        while(p != null) {
            ancestorsOfP.add(p);
            p = childParentMap.get(p);
        }
        
        while(!ancestorsOfP.contains(q)) {
            q = childParentMap.get(q);
        }
        
        return q;
    }
    
    /*public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {       
        if(root == null ) { return null; }
        
        if(root == p || root == q) { return root; }
        
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if(left != null && right != null) {
            return root;
        }
        
        return left == null? right : left;
    }*/    
}

File: 236_Lowest_Common_Ancestor_of_a_Binary_Tree.py
# 
I traverse the tree bottom-up. If a node is one of the targets, I return it.  
I check both subtrees.  
If both sides find a target, the current node is the lowest common ancestor.  
Otherwise, I propagate the found node upward.  

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        # Base case: hit null or find p or q
        if not root or root == p or root == q:
            return root
        
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        
        # If p and q found in both sides, this root is LCA
        if left and right:
            return root
        
        # Otherwise, return the non-null side
        return left if left else right
